components/accelerator/nvidia/error/xid/store/events.go

```diff
@@ -0,0 +1,299 @@
+package store
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/leptonai/gpud/components"
+	"github.com/leptonai/gpud/components/common"
+	"github.com/leptonai/gpud/log"
+	"github.com/leptonai/gpud/pkg/sqlite"
+)
+
+const (
+	ColumnID              = "id"
+	ColumnTimestamp       = "timestamp"
+	ColumnName            = "name"
+	ColumnType            = "type"
+	ColumnMessage         = "message"
+	ColumnExtraInfo       = "extra_info"
+	ColumnSuggestedAction = "suggested_action"
+
+	DefaultRetentionPeriod = 3 * 24 * time.Hour
+)
+
+type eventKey struct {
+	timestamp time.Time
+	name      string
+	eventType string
+}
+
+type eventCache struct {
+	sync.RWMutex
+	events    map[eventKey]components.Event
+	timeIndex []time.Time
+	dirty     bool
+}
+
+func newEventCache() *eventCache {
+	return &eventCache{
+		events: make(map[eventKey]components.Event),
+	}
+}
+
+func (c *eventCache) add(event components.Event) bool {
+	c.Lock()
+	defer c.Unlock()
+
+	key := eventKey{
+		timestamp: event.Time.Time,
+		name:      event.Name,
+		eventType: string(event.Type),
+	}
+
+	if _, exists := c.events[key]; exists {
+		return false
+	}
+
+	c.events[key] = event
+	c.dirty = true
+	return true
+}
+
+func (c *eventCache) rebuildTimeIndex() {
+	if !c.dirty {
+		return
+	}
+
+	c.timeIndex = make([]time.Time, 0, len(c.events))
+	for k := range c.events {
+		c.timeIndex = append(c.timeIndex, k.timestamp)
+	}
+
+	sort.Slice(c.timeIndex, func(i, j int) bool {
+		return c.timeIndex[i].Before(c.timeIndex[j])
+	})
```

Comment:
there is a race condition in these codes. if every time we need to rebuild time index by events,  so don't need to cache the index result.